### ISSUE TITLE
Lookup type mappers by Type

### ISSUE THREAD

**Issue by DanielThomas** [Netflix/hollow#353]
Noticed this pops up a lot while populating - definitely some safepoint bias, but seems overly expensive to generate the type name every time:

```
"main" #1 prio=5 os_prio=31 tid=0x00007f8650041000 nid=0x2703 runnable [0x000070000b083000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.Object.hashCode(Native Method)
	at java.util.HashMap.hash(HashMap.java:339)
	at java.util.LinkedHashMap.get(LinkedHashMap.java:440)
	at java.lang.Class.getAnnotation(Class.java:3415)
	at com.netflix.hollow.core.write.objectmapper.HollowTypeMapper.getDefaultTypeName(HollowTypeMapper.java:61)
	at com.netflix.hollow.core.write.objectmapper.HollowObjectMapper.getTypeMapper(HollowObjectMapper.java:115)
	at com.netflix.hollow.core.write.objectmapper.HollowObjectMapper.getTypeMapper(HollowObjectMapper.java:111)
	at com.netflix.hollow.core.write.objectmapper.HollowObjectMapper.add(HollowObjectMapper.java:69)
	at com.netflix.hollow.api.producer.CloseableWriteState.add(CloseableWriteState.java:48)
```

**Comment by PaulSandoz** [comment]
Unfortunately it is not possible to leverage the `java.lang.reflect.Type` as the key to the `HollowTypeMapper`.  The translation from `Type` to `String` is lossy as `Class.getSimpleName` is used.  Incidentally this means that two POJOs in different packages but the same simple name will clash as the hollow schema does not have a notion of fully qualified names.  Therefore it is not possible use a static `Map`. Another `ConcurrentMap<Type, String>` instance on  `HollowTypeMapper` is required. Fortunately only one instance of `HollowObjectMapper` is created per `HollowProducer`.

**Comment by PaulSandoz** [comment]
Fixed #357
